### DISCUSSION TITLE
Using `setQueryData` with Playwright cause circular reference

### DISCUSSION THREAD

**Discussion by yu-nobara** [TanStack/query#4026] [Q&A] [answered]
Hello,

I have questions about react-query with playwright.

# Context

I'm using Playwright component test for testing react components.

[Experimental: components | Playwright](https://playwright.dev/docs/test-components)

But, I wrote below codes and run test so `undefined: Maximum call stack size exceeded` error occurred.

```tsx
import MyComponet "./my-component";
import { test, expect } from "@playwright/experimental-ct-react";
import { QueryClientProvider, QueryClient } from "react-query";

test("test sample", async ({ mount, page }) => {
  const queryClient = new QueryClient();

  queryClient.setQueryData("key", { username: "foo"});

  const component = await mount(
    <QueryClientProvider client={queryClient}>
      <MyComponent />
    </QueryClientProvider>
  );
});
```

I investigated react-query code and I found out that `QueryCache` object will be created circular reference when `setQueryData` is called.

https://github.com/TanStack/query/blob/52acd7892969fc726f10d4141f63ee7f0c6ae4c2/packages/query-core/src/queryCache.ts#L101-L111

On the other hand, when `mount` is called in Playwright code, `queryClient` object which has circular reference will pass to `wrapFunctions` finally.

https://github.com/microsoft/playwright/blob/3dc1920ce83b56a408d84e2d43e07ec8976d924d/packages/playwright-test/src/mount.ts#L117-L128

This function reference object recursively, so `undefined: Maximum call stack size exceeded` is occurred.

As workaround, I create wrapper component and import it, so well done.

```tsx
import { QueryClientProvider, QueryClient } from "react-query";

export default function QueryClientProviderWithData({ children }: { children: JSX.Element }) {
  const queryClient = new QueryClient();
  queryClient.setQueryData("key", { username: "foo"});

  return (
    <QueryClientProvider client={queryClient}>
      { children }
    </QueryClientProvider>
  )
}
```

```tsx
import MyComponet "./my-component";
import { test, expect } from "@playwright/experimental-ct-react";
import QueryClientProviderWithData from "./query-client-provider"

test("test sample", async ({ mount, page }) => {
  const component = await mount(
    <QueryClientProviderWithData>
      <MyComponent />
    </QueryClientProviderWithData>
  );
});
```

# Question

1. Do you have any other workaround?
2. Do you have plan to remove circular reference in `QueryCache` object?

Thank you,

**Comment by TkDodo** [comment, marked as answer]
> I investigated react-query code and I found out that QueryCache object will be created circular reference when setQueryData is called.

I don't see the circular reference. All that setQueryData does is to create an entry in the cache...

your component looks fine if you only mount the component once, which is true for testing. Otherwise, I'd put the queryClient in state:

```
const [queryClient] = React.useState(() => {
  const client = new QueryClient()
  client.setQueryData(...)
  return client
})
```

**Reply by nishitaniyuki** [reply]
> All that setQueryData does is to create an entry in the cache...

- [The entry has a reference to the cache](https://github.com/TanStack/query/blob/c08c0a1cd304a7e42e792d1d91fb3da7a84b4704/packages/query-core/src/queryCache.ts#L102)(`Query.cache`).
- And the cache has a reference to the entry(`QueryCache.queries`).

Both of these are private members in terms of TS, but that seems not the case for the generated JS.
And this results in a circular reference at runtime.

Do you have any plan to replace TypeScript private keywords with ECMAScript private fields?

**Reply by TkDodo** [reply]
> Do you have any plan to replace TypeScript private keywords with ECMAScript private fields?

I've looked into this once, especially because private class fields can be safely minified and thus reduce bundle size. However, the browsers we are currently targeting (https://tanstack.com/query/v4/docs/installation#requirements) do not all support private class fields (https://caniuse.com/mdn-javascript_classes_private_class_fields). According to [this blog post](https://leoneck.de/blog/js-private-bundle-size), the runtime overhead added is probably not worth it.

If this request came before we released v4, we could've considered it :)

**Reply by nishitaniyuki** [reply]
Thanks for a reply. Then looking forward to the v5 :)
